packages/vitest/src/node/watch-filter.ts

```diff
@@ -0,0 +1,160 @@
+          onSubmit(undefined)
+          break
+        case key?.name === 'enter':
+        case key?.name === 'return':
+          onSubmit(this.results[this.selectionIndex] || this.currentKeyword || '')
+          this.currentKeyword = undefined
+          break
+        case key?.name === 'up':
+          if (this.selectionIndex && this.selectionIndex > 0)
+            this.selectionIndex--
+          else
+            this.selectionIndex = -1
+
+          break
+        case key?.name === 'down':
+          if (this.selectionIndex < this.results.length - 1)
+            this.selectionIndex++
+          else if (this.selectionIndex >= this.results.length - 1)
+            this.selectionIndex = this.results.length - 1
+
+          break
+        case !key?.ctrl && !key?.meta:
+          if (this.currentKeyword === undefined)
+            this.currentKeyword = str
+
+          else
+            this.currentKeyword += str
+          break
+      }
+
+      if (this.currentKeyword)
+        this.results = await filterFunc(this.currentKeyword)
+
+      this.render()
+    }
+  }
+
+  private render() {
+    this.printKeyword()
+    if (!this.currentKeyword) {
+      this.eraseAndPrint('\nPlease input filter pattern')
+    }
+    else if (this.currentKeyword && this.results.length === 0) {
+      this.eraseAndPrint(`\nPattern matches no results`)
+    }
+    else {
+      const resultCountLine = this.results.length === 1 ? `Pattern matches ${this.results.length} result` : `Pattern matches ${this.results.length} results`
+
+      let resultBody = ''
+
+      if (this.results.length > MAX_RESULT_COUNT) {
+        const offset = this.selectionIndex > SELECTION_MAX_INDEX ? this.selectionIndex - SELECTION_MAX_INDEX : 0
+        const displayResults = this.results.slice(offset, MAX_RESULT_COUNT + offset)
+        const remainingResultCount = this.results.length - offset - displayResults.length
+
+        resultBody = `${displayResults.map((result, index) => (index + offset === this.selectionIndex) ? c.green(` › ${result}`) : c.dim(` › ${result}`)).join('\n')}`
+        if (remainingResultCount > 0)
+          resultBody += '\n' + `${c.dim(`   ...and ${remainingResultCount} more ${remainingResultCount === 1 ? 'result' : 'results'}`)}`
+      }
+      else {
+        resultBody = this.results.map((result, index) => (index === this.selectionIndex) ? c.green(` › ${result}`) : c.dim(` › ${result}`))
+          .join('\n')
+      }
+
+      this.eraseAndPrint(`\n${resultCountLine}\n${resultBody}`)
+    }
+    this.restoreCursor()
+  }
+
+  private keywordOffset() {
+    return `? ${this.message} › `.length + 1
+  }
+
+  private printKeyword() {
+    // move corsor to keyword offset
+    stdout().write(`${ESC}${this.keywordOffset()}G`)
+
+    // erase to the end of line
+    stdout().write(`${ESC}K`)
+
+    stdout().write(this.currentKeyword || '')
+  }
+
+  private eraseAndPrint(str: string) {
+    const lineBreasks = str.split('\n').length - 1
```

Comment:
It's probably a minor edge case, but this line counting might need to take into account terminal width via `process.stdout.columns`.
`prompts` does something like this https://github.com/terkelg/prompts/blob/735603af7c7990ac9efcfba6146967a7dbb15f50/lib/util/clear.js#L18

Reply:
@hi-ogawa  
Thank you for letting me know!
I've fixed this. ([c2245bd](https://github.com/vitest-dev/vitest/pull/4733/commits/c2245bd986d76d0ba5ce271c0cb1e108f06edb35))


https://github.com/vitest-dev/vitest/assets/19663115/08f3350c-e50f-4f1a-b771-cdac12603a50